Update README to Improve Usability and Reproducibility of Code Base (#74)
m2lines/convection-parameterization-in-CAM: README.md

The work is described in a [GRL paper](https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2020GL091363) with [accompanying code available](https://github.com/yaniyuval/Neural_nework_parameterization/tree/v.1.0.3). The repository contains the neural net and its implementation into a simple system for atmospheric modelling, [SAM](http://rossby.msrc.sunysb.edu/~marat/SAM.html).

The aims of this repository are to:
1. provide a standalone fortran module of the convection parameterisation based on this neural net that can be used elsewhere,
2. deploy the module in another atmospheric model, and

> 3. evaluate its performance (and provide instructions/code to do this).

Reviewer: I guess this is now done in https://github.com/m2lines/CAM-ML.
Initially we were advised that all the work would be done here, but later it was suggested that we use a fork of CAM to implement it. So perhaps this needs changing now.
Though evaluation results might be reported on this repo once we finalise them.

Author: Do you have any suggestion on how to change the text here? Or leave it like this?

Reviewer: Not really...
I think leave for now perhaps.
I think the CAM-ML repo is linked from elsewhere in the README.

> We have also started to perform an investigation into interfacing the pytorch implementation of the Neural Net using the [pytorch-fortran bridging code](https://github.com/Cambridge-ICCS/fortran-pytorch-lib) developed at the [Institute of Computing for Climate Science](https://cambridge-iccs.github.io/).

Reviewer: I wonder if we just remove this for now?
No longer a focus for M2LInES and we could add it back later if we do something?

Author: But there is still some leftover code/NN weights?

Reviewer: Not in the main repository that I can see, it's all in unmerged and stale (and messy) pull requests I think.

The model was first deployed into the [Single Column Atmospheric Model (SCAM)](https://www.cesm.ucar.edu/models/simple/scam) - a single column version of the CESM.

> We evaluate performance using Single Column CAM (SCAM) in the gateIII configuration for tropical convection in a similar manner described by the [SCAM6 pulication in JAMES](https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2018MS001578).

Reviewer: More a note-to-self, we are going to move to focus on comparing to the TOGA II dataset which is a similar idea to GATE III run in SCAM. Documented in the same paper we already cite here.

Author: Do we want to mention it here?

Reviewer: Yes, I think maybe mention both GATE III and TOGA II?

Author: Ah, just seen that you said it's the same paper

Author: added that to the readme

This compares model performance to data from an intense observation period (IOP) described in an [AMS publication](https://journals.ametsoc.org/view/journals/atsc/36/1/1520-0469_1979_036_0053_saposs_2_0_co_2.xml).

Long term developments of this project will seek to re-deploy more complex ML parameterizations into mode complex atmospheric models such as the [Community Atmospheric Model (CAM)](https://www.cesm.ucar.edu/models/cam) part of the CESM.

## Repository structure

```
├── YOG_convection
│   └── ...
├── torch_nets
│   └── ...

> ├── CAM_interface

Reviewer: We now have `profile_test` as of this morning.

Author: added that

It has dependencies on the NetCDF and NetCDF-Fortran libraries and can be built using the included makefile (may need modification).

Files:

- `nn_cf_net.f90` - The core neural net routines
- `nn_convection_flux.f90` - wrapper around the neural net routines to perform physics-based operations
- `test.f90` - simple smoke tests for parameterisation routines
- `NN_weights_YOG_convection.nc` - NetCDF file with weights for the neural net
- Makefile - Makefile to compile these files

Note that you will need to generate the SAM sounding (an [atmospheric sounding](https://en.wikipedia.org/wiki/Atmospheric_sounding) that defines the SAM grid that the parameterisation is operating on -- 
we need this grid for the parameterisation, but also to interpolate any inputs from a different model on to this grid so that they can be used in the neural net) as a NetCDF file from the data files in `resources/` if you are using the CAM interface.\
This can be done from within the `resources/` directory as follows:
```
python -m venv venv
source venv/bin/activate
pip install -r requirements.txt
python sounding_to_netcdf.py
deactivate

```

> There are test routines associated with this code in `/tests/test_NN_module/`.

Reviewer: ```suggestion
There are test routines associated with this code in `/tests/test_YOG_convection/`.
```

There are notable future changes to the CAM source for CESM v2.2 and v3.0 that may
require revisions to these files in future.

In addition to the files in `[/YOG_convection](/YOG_convection)` containing the parameterisation,
the additional files here are required:

- `nn_interface_CAM.F90` - The interface for performing conversion from CAM variables and grid into the variables/grid expected by the YOG parameterisation.
- `yog_intr.F90` - The interface between the CAM model and the YOG parameterisation.

Changes also have to be made to:

- `bld/namelist_files/namelist_definition.xml` - To add the new namelist variables.
- `bld/namelist_files/namelist_defaults_cam.xml` - To set the new namelist variable defaults.
- `/src/physics/cam/physcontrol.F90` - To read in the new namelist variables and broadcast them.
- `/src/physics/cam/physpkg.F90` - To call the new parameterisation routines detailed in `yog_intr.F90`.

Full details of these can be seen in the specific implementation in CAM linked above.

There are test routines associated with this code in `/tests/test_CAM_interface/`.

> Guidance on running these can be found in the main README.

Reviewer: ```suggestion
Guidance on running these can be found below.
```
